### PR TITLE
Add some BinarySearch tests to Array

### PR DESCRIPTION
**Pull request by hughbe**
- All primitive types
- Empty array
- Tests for when the array has a non-zero lower bound
- More tests for invalid index + length
- More tests for invalid types
- Tests for when count is zero
- Cleans up valid and invalid test inputs for BinarySearch. This means
that coverage is increased for both the generic and non-generic
overloads.

/cc @stephentoub

### CODE REVIEW

File: src/System.Runtime/tests/System/ArrayTests.cs [hunk 1]
```diff
@@ -129,49 +130,239 @@ public static void Construction_MultiDimensionalArray()
             Assert.Equal(array.GetValue(0, 0, 0, 2), 3);
         }
 
-        public static IEnumerable<object[]> BinarySearch_NonGeneric_TestData()
+        public static IEnumerable<object[]> BinarySearch_TestData()
         {
             string[] stringArray = new string[] { null, "aa", "bb", "bb", "cc", "dd", "ee" };
+            string stringDummy = "";
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/ArrayTests.cs]: Why are all these dummy values necessary?  They clutter up the tests and I'd prefer they need be added.  Are these all just here for generic inference?  There's no better way to achieve this?

Comment (hughbe) [inline src/System.Runtime/tests/System/ArrayTests.cs]: Yeah - its for generic inference. I actually made a PR to xunit yesterday to fix this: https://github.com/xunit/xunit/pull/965

Comment (stephentoub) [inline src/System.Runtime/tests/System/ArrayTests.cs]: But this is only an issue because ~10 of the entries you added try to pass null as the value, which forced you to change the parameter to be ```object value```, which meant you needed to add a dummy additional parameter of type T.  If you just got rid of those ~10 entries, then you could make the parameter ```T value```, and not need the additional dummy parameter.  Almost all of those entries don't make sense for the generic version, anyway, as null isn't a valid value for those primitive types, so doing this would also let you remove the special case that bails early in the case of null, remove the now unnecessary casts from object to T, etc.  You can then add just those ~10 entries as data input to the non generic test, where null can make sense.

Comment (hughbe) [inline src/System.Runtime/tests/System/ArrayTests.cs]: Good point!

Comment (hughbe) [inline src/System.Runtime/tests/System/ArrayTests.cs]: Fixed

File: src/System.Runtime/tests/System/ArrayTests.cs [hunk 2]
```diff
@@ -192,36 +368,53 @@ public static void BinarySearch_NonGeneric(Array array, int index, int length, o
...
             }
             if (isDefaultComparer)
             {
                 // Use BinarySearch<T>(T, int, int, T)
-                Assert.Equal(expected, Array.BinarySearch(array, index, length, value));
+                Assert.Equal(expected, Array.BinarySearch(array, index, length, (T)value));
             }
             // Use BinarySearch<T>(T[], int, int, T, IComparer)
-            Assert.Equal(expected, Array.BinarySearch(array, index, length, value, comparer));
+            Assert.Equal(expected, Array.BinarySearch(array, index, length, (T)value, comparer));
```
Comment (stephentoub) [inline src/System.Runtime/tests/System/ArrayTests.cs]: Why are all of these (T) casts necessary? value is already T.

**Comment by stephentoub** [comment]
Thanks, @hughbe.
